### PR TITLE
[blog] Written a blog in Go series

### PR DESCRIPTION
**Pull request by grgprarup**
- Added a blog on `go` fundamentals in `Journey With Go - A Blog Series`

### CODE REVIEW

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 1]
```diff
@@ -0,0 +1,821 @@
...
+tags: function, array, rune, method, interface, string, struct
+banner: https://blog.jankaritech.com/src/assets/JourneyWithGo-ABlogSeries/images/control-statements-banner.svg
+seriesTitle: Journey With Go - A Blog Series
+episode: 3
+---
+
+**[गो](https://go.dev "Go in Nepali")** **"Journey With Go - A Blog Series"** about the basics of Go `Go Basic`,
+advanced concepts in Go `Go Beyond Basics`, testing with go `Godog as Test Framework` and many more.
+
+This is the third installment of the series. You are here which means you have already read the first and second blog of
```
Comment (koebel) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:16]: There might be some people who haven't read the previous parts of the series, therefore I would use a modifier like 'most likely' or similar: 

_You are here which means you have most likely already read the first and second blog of the series._

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 2]
```diff
@@ -0,0 +1,821 @@
...
+It is particularly important when dealing with text in different languages or when working with special symbols and
+emojis. It ensures that each character is treated as a single entity, regardless of its complexity or language.
+
+## **2. String**
+
+In Go, a string is a sequence of bytes (not characters) that is used to represent text. Strings are immutable, which
+means once created, their values cannot be changed. However, you can assign a new value to a string variable, which will
+create a new string.
+
+```go
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:62]: A Note stating the meaning of `' '` (rune) and `" "` (string) would be great here. like using `'` is a rune and using `"` is a string.

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 3]
```diff
@@ -0,0 +1,821 @@
...
+  fmt.Println("Hello, " + "World!") // Hello, World!, + can be used to concatenate strings
+}
+```
+
+### **2.1. Slicing of String**
+
+Slicing is a mechanism to extract a portion of a string. It is done by specifying the start and end indices of the part
+of the string that you want to extract. The syntax for slicing a string is `string[start:end]`.
+
+| slice            | output        | description                                                                                    |
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:86]:
```suggestion
of the string that you want to extract. The syntax for slicing a string is `string[start:end]`.

Taking the above `greeting` variable as example:

| slice            | output        | description                                                                                    |
```

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 4]
```diff
@@ -0,0 +1,821 @@
...
+Index: 6, Value:
+Index: 7, Value: W
+Index: 8, Value: o
+Index: 9, Value: r
+Index: 10, Value: l
+Index: 11, Value: d
+Index: 12, Value: !
+```
+
+### **2.3. Rune Count Vs Len**
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:128]: if we are comparing different methods
```suggestion
### **2.3. `len` vs `RuneCountInString`**
```

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 5]
```diff
@@ -0,0 +1,821 @@
...
+}
+```
+
+In the example for the English language, the value returned by the `len` function is same as of the value returned by the `utf8.RuneCountInString` function.
+
+However, for the Thai language, the value returned by the `len` function is different from the value returned by the `utf8.RuneCountInString` function. This is because the Thai language uses the UTF-8 encoding scheme, which uses 3 bytes to represent a single character. Therefore, the string "สวัสดี" has 18 bytes, but only 6 runes.
+
+## **3. Array**
+
+In Go, an array is a fixed-length sequence of elements of a single type. Arrays are useful when you know the number of elements that you want to store in advance.
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:159]:
```suggestion
In Go, an array is a fixed-length sequence of elements of the same type. Arrays are useful when you know the number of elements that you want to store in advance.
```

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+
+// defining a method named `area` that takes no parameters and returns a float64
+func (t Triangle) area() float64 {
+  return 0.5 * t.base * t.height
+}
+
+func area() float64 {
+	t := Triangle{base: 10, height: 5}
+	return 0.5 * t.base * t.height
+}
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:603]: hmm? I thinks it's not required here
```suggestion
```

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 7]
```diff
@@ -0,0 +1,825 @@
...
+
+  // changing the value of a field
+  student.name = "Jane Doe"
+  fmt.Println(student.name) // Jane Doe
+}
+```
+
+## **6. Method**
+
+In Go, a method is a function that is associated with a type. It is useful for grouping related functions together. Methods are useful when you want to model real-world entities that have multiple behaviors. Methods are also known as functions with receivers.
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:573]: `struct` would be easier to understand?
```suggestion
In Go, a method is a function that is associated with a struct. It is useful for grouping related functions together. Methods are useful when you want to model real-world entities that have multiple behaviors. Methods are also known as functions with receivers.
```

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 8]
```diff
@@ -0,0 +1,825 @@
...
+```
+
+## What we have learned so far
+
+- Rune and String
+- Array
+- Function and Method
+- Struct and Interface
+
+## Conclusion
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:805]: I kinda feel like repetition. Maybe only `Conclusion` section is enough?

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 9]
```diff
@@ -0,0 +1,825 @@
...
+  fmt.Println(triangle) // {10 5}
+
+  area := triangle.area()
+  fmt.Println("Area of triangle is " + fmt.Sprintf("%.2f", area)) // 25
+
+  perimeter := triangle.perimeter()
+  fmt.Println("Perimeter of triangle is " + fmt.Sprintf("%.2f", perimeter)) // 25
+
+  rectangle := Rectangle{length: 10, width: 5}
+  fmt.Println(rectangle) // {10 5}
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:778]: I didn't see use of `Shape` interface. maybe use it something like this
```suggestion
func main() {
  var triangle Shape = Triangle{base: 10, height: 5}
  fmt.Println(triangle) // {10 5}

  area := triangle.area()
  fmt.Println("Area of triangle is " + fmt.Sprintf("%.2f", area)) // 25

  perimeter := triangle.perimeter()
  fmt.Println("Perimeter of triangle is " + fmt.Sprintf("%.2f", perimeter)) // 25

  var rectangle Shape = Rectangle{length: 10, width: 5}
  fmt.Println(rectangle) // {10 5}
```

Comment (saw-jan) [reply src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:778]: or some other usage

File: src/assets/JourneyWithGo-ABlogSeries/goBasics.md [hunk 10]
```diff
@@ -1,11 +1,11 @@
...
 title: Go Basics - The Starting Point
 authorName: Prarup Gurung (प्रारुप गुरुङ्ग)
-authorAvatar: https://www.jankaritech.com/images/2021/12/29/p1110938-crop-colour-800.jpg
+authorAvatar: https://avatars.githubusercontent.com/u/8559121?v=4
 authorLink: https://github.com/grgprarup
 createdAt: Dec 6, 2022
-tags: golang, introduction, go variables, go datatypes, constants
-banner: https://camo.githubusercontent.com/2b507540e2681c1a25698f246b9dca69c30548ed66a7323075b0224cbb1bf058/68747470733a2f2f676f6c616e672e6f72672f646f632f676f706865722f6669766579656172732e6a7067
+tags: golang, go, introduction, go variables, go datatypes, constants
+banner: https://blog.jankaritech.com/src/assets/JourneyWithGo-ABlogSeries/images/go-basics-banner.jpeg
```
Comment (saw-jan) [inline src/assets/JourneyWithGo-ABlogSeries/goBasics.md:8]: if this can be in next PRs

**Comment by koebel** [comment]
This is maybe a more general comment regarding our blog posts: we should make sure to state the source of the header images unless they are our own creations or come from a source that clearly states that no attribution to the creator is requested...

**Comment by grgprarup** [comment]
> This is maybe a more general comment regarding our blog posts: we should make sure to state the source of the header images unless they are our own creations or come from a source that clearly states that no attribution to the creator is requested...

The source and credits to the creator of any external images go to the end of my blog.

**Comment by koebel** [comment]
> > This is maybe a more general comment regarding our blog posts: we should make sure to state the source of the header images unless they are our own creations or come from a source that clearly states that no attribution to the creator is requested...
> 
> The source and credits to the creator of any external images go to the end of my blog.

perfect

File: src/components/detail/ContentData.vue [hunk 11]
```diff
@@ -68,11 +68,10 @@ defineProps({
   pre {
     white-space: pre-wrap;
     background-color: rgb(225 225 225);
-    overflow-x: auto;
     padding: 0.7rem 1rem;
     max-height: 400px;
     width: 100%;
-    overflow-y: auto;
+    overflow: auto;
```
Comment (saw-jan) [inline src/components/detail/ContentData.vue:74]: blog looks good to me. unrelated changes like these could be another small PR?

**Review by saw-jan — COMMENTED** [review]
:sparkles: from my side

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 12]
```diff
@@ -0,0 +1,821 @@
...
+Index: 1, Value: Javascript
+Index: 2, Value: Python
+```
+
+## **4. Function**
+
+In Go, a function is a block of code that performs a specific task. Functions are useful when you want to reuse the same code multiple times. They also help in organizing your code into smaller chunks, which makes it easier to read and maintain.
+
+### **4.1. Function Definition**
+A function must be defined before it can be called. A function definition must have a name, a list of parameters, a return type, and a body. In Go, a `func` keyword is used to define a function.
```
[the remark this answers is not in the document]
Comment (koebel) [reply src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:346]: @phil-davis correct me if this is not correct

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 13]
```diff
@@ -0,0 +1,821 @@
...
+
+### **2.2. Iterating over String**
+
+Strings are immutable, but you can iterate over them using a for loop with range.
+
+```go
+package main
+
+import (
+  "fmt"
```
[the remark this answers is not in the document]
Comment (grgprarup) [reply src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:105]: The `fmt` is already used in the first blog. So I omit it here.

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 14]
```diff
@@ -0,0 +1,821 @@
...
+)
+
+func main() {
+  greeting := "Hello, World!"
```
[the remark this answers is not in the document]
Comment (grgprarup) [reply src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:109]: This is explained in the first blog.

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 15]
```diff
@@ -0,0 +1,821 @@
...
+func swap(a int, b int) (int, int) {
+  return b, a
+}
+
+func main() {
+  greet("World") // Calling the `greet` function
+  fmt.Println(add(1, 2)) // Calling the `add` function
+
+  sum := add(3,6) // Calling the `add` function and storing the returned value in a variable `sum`
+  fmt.Println(sum)
```
[the remark this answers is not in the document]
Comment (grgprarup) [reply src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:463]: I used the output section here.

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 16]
```diff
@@ -0,0 +1,821 @@
...
+}
+
+// defining a function named `swap` that takes two integer parameters and returns two integers
+func swap(a int, b int) (int, int) {
+  return b, a
+}
+
+func main() {
+  greet("World") // Calling the `greet` function
+  fmt.Println(add(1, 2)) // Calling the `add` function
```
[the remark this answers is not in the document]
Comment (grgprarup) [reply src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:460]: I have changed the use case of the example.

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 17]
```diff
@@ -0,0 +1,972 @@
...
+
+  triangle := Triangle{base: 10, height: 5}                         // creating a new instance of the `Triangle` struct
+  fmt.Println(triangle)
+
+  // calling the `area` method
+  area = triangle.getAreaOfTriangle()                               // triangle is the receiver
+  fmt.Println("Area of triangle is " + fmt.Sprintf("%.2f", area))
+}
+```
+
```
Comment (koebel) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:823]: make sure that the function names in the comments match with the actual function names

File: src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md [hunk 18]
```diff
@@ -0,0 +1,972 @@
...
+- String
+- Array
+- Function
+- Struct
+- Method
+- Interface
+
+In the next blog, we will learn about some advanced topics like Errors, Goroutines, and Channels in Go language.
+
+*Keep learning and keep practicing*
```
Comment (koebel) [inline src/assets/JourneyWithGo-ABlogSeries/goFundamentals.md:960]: add point or exclamation mark at the end of the sentence

**Comment by koebel** [comment]
Looks good, I like your new example for Runes with Devanagari Script!

**Review by saw-jan — APPROVED** [review]
:sparkles:
